compiler/front/sexp_reporter.nim

```diff
@@ -0,0 +1,185 @@
+## Implementation of the structured CLI message generator. Using
+## `--msgFormat=sexp` will make compiler switch to the report hook
+## implemented in this module.
+
+import
+  experimental/[
+    sexp,
+    diff,
+    colortext,
+    sexp_diff
+  ],
+  ast/[
+    lineinfos,
+    ast,
+    reports
+  ],
+  front/[
+    options,
+    msgs
+  ],
+  std/[
+    strutils
+  ]
+
+
+var writeConf: ConfigRef
+
+import std/options as std_options
+
+proc addFields[T](s: var SexpNode, r: T, ignore: seq[string] = @[])
+
+proc sexp[T: object | tuple](obj: T): SexpNode =
+  result = newSList()
+  addFields(result, obj)
+
+proc sexp[T: object | tuple](obj: ref T): SexpNode =
+  result = newSList()
+  addFields(result, obj[])
+
+proc sexp*[E: enum](e: E): SexpNode = newSSymbol($e)
+
+proc sexpItems*[T](s: T): SexpNode =
+  result = newSList()
+  for item in items(s):
+    result.add sexp(item)
+
+
+proc sexp*[T](s: seq[T]): SexpNode = sexpItems(s)
+proc sexp*[R, T](s: array[R, T]): SexpNode = sexpItems(s)
+proc sexp*[I](s: set[I]): SexpNode = sexpItems(s)
+proc sexp*(s: cstring): SexpNode = sexp($s)
+
+proc sexp*(v: SomeInteger): SexpNode = newSInt(BiggestInt(v))
+proc sexp*(id: FileIndex): SexpNode =
+  sexp(writeConf.toMsgFilename(id))
+
+
+iterator sexpFields[T](obj: T, ignore: seq[string] = @[]): SexpNode =
```

Comment:
Why does it feel like a bug? Generic iterator, accepts generic type `T`, works as intended

> I think T should have an a type constraint of object | tuple

Can be clarified later, but the fact it works without it doesn't really make it a bug IMO

Reply:
It absolutely is a bug because there is no guarantee you can iterator over T?

Reply:
What do you mean "can't iterator over T"? The `iterator` is defined for `T` therefore I can iterate over it, I don't see any bugs here

Reply:
`T` is not guaranteed to have fields for `fieldPairs` to work, it might be out of scope for this PR, but it's silly that between the stdlib and the compiler that this is allowed.

Reply:
You are talking about need for 'HasFields' concept or something like that here, otherwise it is a perfectly fine generic that would generate a perfectly fine compilation error on instantiation saying that there are no fields in here. I don't understand where the "bug" part comes from here. Just like the sexpItems above - it is instantiation error, just like every other generic proc. I can put object|tuple there, but that would just make code clearer, it won't make it more correct (as in - reject malformed program that would've compiled previously) in any way. If there is no field pairs the code does not compile, pointing to "instantiation of X...". If there are field pairs the code compiles.